src/platforms/atomic-kms/server/kms/cursor.cpp

```diff
@@ -0,0 +1,438 @@
+/*
+ * Copyright © Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License version 2 or 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "cursor.h"
+#include "platform.h"
+#include "kms_output.h"
+#include "kms_output_container.h"
+#include "kms_display_configuration.h"
+#include "mir/geometry/rectangle.h"
+#include "mir/graphics/cursor_image.h"
+
+#include <xf86drm.h>
+
+#include <boost/exception/errinfo_errno.hpp>
+
+#include <stdexcept>
+#include <vector>
+
+namespace mg = mir::graphics;
+namespace mga = mg::atomic;
+namespace geom = mir::geometry;
+
+namespace
+{
+// Transforms a relative position within the display bounds described by \a rect which is rotated with \a orientation
+geom::Displacement transform(geom::Rectangle const& rect, geom::Displacement const& vector, MirOrientation orientation)
+{
+    switch(orientation)
+    {
+    case mir_orientation_left:
+        return {vector.dy.as_int(), rect.size.width.as_int() -vector.dx.as_int()};
+    case mir_orientation_inverted:
+        return {rect.size.width.as_int() -vector.dx.as_int(), rect.size.height.as_int() - vector.dy.as_int()};
+    case mir_orientation_right:
+        return {rect.size.height.as_int() -vector.dy.as_int(), vector.dx.as_int()};
+    default:
+    case mir_orientation_normal:
+        return vector;
+    }
+}
+
+int get_drm_cursor_height(int fd)
+{
+    uint64_t height;
+    drmGetCap(fd, DRM_CAP_CURSOR_HEIGHT, &height);
+    return int(height);
```

Comment:
I know this is pre-existing, but this is an unusual way of spelling
```suggestion
    return static_cast<int>(height);
```